### PR TITLE
CC-3615 add remove-ip and set-ip commands

### PR DESCRIPTION
**Pull request by ssOlexBaiko**
Fix provides two additional commands to cli package to remove and set new IP Assignment.
[Jira](https://jira.zenoss.com/browse/CC-3615)

### CODE REVIEW

File: cli/api/service.go [hunk 1]
```diff
@@ -399,6 +401,42 @@ func (a *api) AssignIP(config IPConfig) error {
 	return nil
 }
 
+// RemoveIP removes IP assignments of a service
+func (a *api) RemoveIP(id string) error {
```
Comment (smousa) [inline cli/api/service.go]: May not want to remove by ID, rather by service id and endpoint

File: cli/cmd/service.go [hunk 2]
```diff
@@ -467,6 +467,30 @@ func (c *ServicedCli) initService() {
 				BashComplete: c.printServicesFirst,
 				Action:       c.cmdServiceClearEmergency,
 			},
+			{
+				Name:         "remove-ip",
+				Usage:        "Remove the IP assignment of a service's endpoints",
+				Description:  "serviced service remove-ip SERVICEID",
```
Comment (smousa) [inline cli/cmd/service.go]: need to specify endpoint as well

File: cli/cmd/service.go [hunk 3]
```diff
@@ -467,6 +467,30 @@ func (c *ServicedCli) initService() {
...
+				BashComplete: c.printServicesFirst,
+				Action:       c.cmdServiceRemoveIP,
+			},
+			{
+				Name:         "set-ip",
+                                Usage:        "Setting an IP address to a service's endpoints requiring an explicit IP address. If ip is not provided it does an automatic assignment",
+                                Description:  "serviced service set-ip SERVICEID [ADDRESS] [--port=PORT] [--proto=PROTOCOL]",
```
Comment (smousa) [inline cli/cmd/service.go]: need to specify endpoint name

**Comment by ssOlexBaiko** [comment]
retest this

File: facade/service.go [hunk 4]
```diff
@@ -720,6 +720,37 @@ func (f *Facade) validateServiceMigration(ctx datastore.Context, svcs []service.
 	return nil
 }
 
+func (f *Facade) RemoveIPs(ctx datastore.Context, args []string) error {
```
Comment (tgjonestx) [inline facade/service.go]: Should have `defer ctx.Metrics().Stop(ctx.Metrics().Start("Facade.RemoveIPs"))`

File: facade/service.go [hunk 5]
```diff
@@ -720,6 +720,37 @@ func (f *Facade) validateServiceMigration(ctx datastore.Context, svcs []service.
...
+			}
+			endpoint.RemoveAssignment()
+		}
+	}
+
+	return nil
+}
+
+func (f *Facade) SetIPs(ctx datastore.Context, request addressassignment.AssignmentRequest) error {
+	defer ctx.Metrics().Stop(ctx.Metrics().Start("Facade.SetIPs"))
```
Comment (tgjonestx) [inline facade/service.go]: This call should be audit logged too

File: facade/service.go [hunk 6]
```diff
@@ -720,6 +720,37 @@ func (f *Facade) validateServiceMigration(ctx datastore.Context, svcs []service.
 	return nil
 }
 
+func (f *Facade) RemoveIPs(ctx datastore.Context, args []string) error {
+	serviceID := args[0]
```
Comment (tgjonestx) [inline facade/service.go]: This call needs to be recorded in the CC audit log. Look through this file for `f.auditLogger` for examples of how to audit log something.

File: domain/service/service.go [hunk 7]
```diff
@@ -803,6 +803,23 @@ func (se *ServiceEndpoint) SetAssignment(aa addressassignment.AddressAssignment)
 	return nil
 }
 
+//SetAddressConfig sets the AddressConfig for the endpoint
+func (s Service) SetAddressConfig(endpointName string, sa servicedefinition.AddressResourceConfig) error {
+	if s.Endpoints == nil {
+		return fmt.Errorf("Service %s has no Endpoints", s.Name)
+	}
+
+	for i := range s.Endpoints {
```
Comment (smousa) [inline domain/service/service.go]: Should return an error if the endpoint does not exist

File: facade/publicendpoint.go [hunk 8]
```diff
@@ -183,6 +183,32 @@ func (f *Facade) RemovePublicEndpointPort(ctx datastore.Context, serviceid, endp
 	return nil
 }
 
+// SetAddressConfig sets the AddressConfig to the endpoint
+func (f *Facade) SetAddressConfig(ctx datastore.Context, serviceid, endpointName string, sa servicedefinition.AddressResourceConfig) error {
+	defer ctx.Metrics().Stop(ctx.Metrics().Start("Facade.SetAddressConfig"))
+	svc, err := f.GetService(ctx, serviceid)
+	if err != nil {
+                err = fmt.Errorf("Could not find service %s: %s", serviceid, err)
+                glog.Error(err)
```
Comment (smousa) [inline facade/publicendpoint.go]: Don't use glog--use logrus instead

Comment (smousa) [inline facade/publicendpoint.go]: Also, you wouldn't wrap your log messages in errors.

File: facade/publicendpoint.go [hunk 9]
```diff
@@ -183,6 +183,32 @@ func (f *Facade) RemovePublicEndpointPort(ctx datastore.Context, serviceid, endp
...
+		glog.Error(err)
+		return err
+	}
+
+	if err = f.UpdateService(ctx, *svc); err != nil {
+		glog.Error(err)
+		return err
+	}
+
+	glog.V(2).Infof("Service (%s) updated", svc.Name)
```
Comment (smousa) [inline facade/publicendpoint.go]: Again, no glog, anywhere

File: facade/service.go [hunk 10]
```diff
@@ -1954,24 +1988,69 @@ func (f *Facade) restoreIPs(ctx datastore.Context, svc *service.Service) error {
...
+func IpVisitorFn(command string, request addressassignment.AssignmentRequest, f *Facade, ctx datastore.Context) service.Visit {
 	visitor := func(svc *service.Service) error {
-		// get all of the ports for the service
-		portmap, err := GetPorts(svc.Endpoints)
-		if err != nil {
-			glog.Errorf("Could not get ports for service %s (%s): %s", svc.Name, svc.ID, err)
-		} else if len(portmap) == 0 {
+		var portmap Ports
+		var auditMessage string
+		if command == "assign-ip" {
```
Comment (smousa) [inline facade/service.go]: Why are you string matching here?
You should have separate functions based on the actions that you are trying to perform.

File: domain/addressassignment/addressassignment.go [hunk 11]
```diff
@@ -36,6 +36,9 @@ type AssignmentRequest struct {
 	ServiceID      string
 	IPAddress      string
 	AutoAssignment bool
+        Port           uint16
```
Comment (smousa) [inline domain/addressassignment/addressassignment.go]: Spacing here

File: facade/publicendpoint.go [hunk 12]
```diff
@@ -183,6 +183,34 @@ func (f *Facade) RemovePublicEndpointPort(ctx datastore.Context, serviceid, endp
...
+
+	alog := f.auditLogger.Action(audit.Update).Message(ctx, "Remove IP Assignment").
+		WithFields(logrus.Fields{
+			"servicename": svc.Name,
+			"endpoint": endpointName,
+		}).Entity(svc)
+
+	err = svc.SetAddressConfig(endpointName, sa)
+	if err != nil {
+		err = fmt.Errorf("Error setting AddressConfig to the service (%s): %v", svc.Name, err)
```
Comment (smousa) [inline facade/publicendpoint.go]: error messages should not be capitalized

File: facade/service.go [hunk 13]
```diff
@@ -2060,8 +2120,39 @@ func (f *Facade) AssignIPs(ctx datastore.Context, request addressassignment.Assi
...
+func AssignIpArgs(ctx datastore.Context, f *Facade, request addressassignment.AssignmentRequest) (IpArgs, error) {
+	var arguments IpArgs
+	auditMessage := "Assign IP Address"
+	svc, err := f.GetService(ctx, request.ServiceID)
+	if err != nil {
+		return arguments, err
+	}
+	portmap, err := GetPorts(svc.Endpoints)
+	if err != nil {
+		glog.Errorf("Could not get ports for service %s (%s): %s", svc.Name, svc.ID, err)
```
Comment (smousa) [inline facade/service.go]: remove glog

**Comment by tgjonestx** [comment]
Build finished.

**Pull request merged**